web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useEffect, useState } from 'react';
+import { useFileTransferContext } from 'shared/components/FileTransfer';
+import { FileTransferRequest } from 'shared/components/FileTransfer/FileTransferRequests';
+
+import Tty from 'teleport/lib/term/tty';
+import { EventType } from 'teleport/lib/term/enums';
+import { Session } from 'teleport/services/session';
+import { DocumentSsh } from 'teleport/Console/stores';
+
+import { useConsoleContext } from '../consoleContextProvider';
+
+import { getHttpFileTransferHandlers } from './httpFileTransferHandlers';
+import useGetScpUrl from './useGetScpUrl';
+
+const useFileTransfer = (
+  tty: Tty,
+  session: Session,
+  currentDoc: DocumentSsh,
+  addMfaToScpUrls: boolean
+) => {
+  const { filesStore } = useFileTransferContext();
+  const ctx = useConsoleContext();
+  const user = ctx.getStoreUser();
+  const [fileTransferRequests, setFileTransferRequests] = useState<
+    FileTransferRequest[]
+  >([]);
+  const { getScpUrl, attempt: getMfaResponseAttempt } =
+    useGetScpUrl(addMfaToScpUrls);
+
+  useEffect(() => {
+    // the tty will be init outside of this hook, so we wait until
+    // it exists and then attach file transfer handlers to it
+    if (tty) {
+      tty.on(EventType.FILE_TRANSFER_REQUEST, updateFileTransferRequests);
+      tty.on(
+        EventType.FILE_TRANSFER_REQUEST_APPROVE,
+        handleFileTransferApproval
+      );
+      tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+    }
+  }, [tty]);
+
+  function removeFileTransferRequest(requestId: string) {
+    setFileTransferRequests(prevstate =>
+      prevstate.filter(ft => ft.requestID !== requestId)
+    );
+  }
+
+  async function download(
+    location: string,
+    abortController: AbortController,
+    moderatedSessionParams?: ModeratedSessionParams
+  ) {
+    if (session.moderated) {
```

Comment:
I don't understand how the `moderated` session works I think :( 

1. User tries to download a file so `download` is invoked.
2. We are in a moderated session so a request is sent and we abort early. 
3. The approval comes in and we run `download` again.
4. `session.moderated` is false now, correct? When does the value change?

Reply:
Just from my shallow understanding of this I think `session.moderated` comes from #24238 and is set per-session. That is, you create a moderated session and it stays "moderated" throughout its lifespan, it cannot become "unmoderated".

Reply:
>That is, you create a moderated session and it stays "moderated" throughout its lifespan, it cannot become "unmoderated".

This makes sense, but I don't understand how the code here  https://github.com/gravitational/teleport/blob/dac00197f876db232a5a5a3841c6dcf6210f9c21/web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts#L139 runs the actual download 🤔

Reply:
If `session.moderated` is true, we only send the `tty.sendFileDownloadRequest`. The actual file transfer is handled when we receive an `approved` message from the server. However, you are correct that this is WRONG! It took me awhile to figure out what happened because it actually works (when you read it, it shouldnt!). 

Originally, when `useFileTransfer` was inside of `useSshSession`, I had a handleUpload, and handleDownload function exported and it did that check. I accidentally conflated those two when doing the refactor and removed it thinking "sweet, I can just have one function now". 

The reason why this 'works' now is the function `download` that we are passing to `handleFileTransferApproval` is passed before `session` is defined and we are inside a memoized document. So `session.moderated` isn't even set in _that version_ of `download`, where as the one the user uses from the file transfer dialog checks session that that invocation time.

So even tho it works, it really shouldn't and is no way readable in it's current state. I'm going to add back the `download`/`handleDownload` paradigm and it'll work as expected and be much more readable. Thanks for calling this out